packages/synchronize/lib/src/sync.dart

```diff
@@ -0,0 +1,246 @@
+import 'package:synchronize/src/action.dart';
+import 'package:synchronize/src/conflict.dart';
+import 'package:synchronize/src/journal.dart';
+import 'package:synchronize/src/journal_entry.dart';
+import 'package:synchronize/src/object.dart';
+import 'package:synchronize/src/sources.dart';
+
+/// Sync between two [SyncSources]s.
+///
+/// This implementation follows https://unterwaditzer.net/2016/sync-algorithm.html in a generic and abstract way
+/// and should work for any two kinds of sources and objects.
+Future<List<SyncConflict<T1, T2>>> sync<T1, T2>(
+  final SyncSources<T1, T2> sources,
+  final SyncJournal journal,
+  final Map<String, SyncConflictSolution> conflictSolutions, {
+  final bool keepSkipsAsConflicts = false,
+}) async {
+  final diff = await computeSyncDiff<T1, T2>(
+    sources,
+    journal,
+    conflictSolutions,
+    keepSkipsAsConflicts: keepSkipsAsConflicts,
+  );
+  await executeSyncDiff<T1, T2>(
+    sources,
+    journal,
+    diff,
+  );
+  return diff.conflicts;
+}
+
+/// Differences between the two sources
+class SyncDiff<T1, T2> {
+  // ignore: public_member_api_docs
+  SyncDiff(
+    this.actions,
+    this.conflicts,
+  );
+
+  /// Actions required to solve the difference
+  final List<SyncAction> actions;
+
+  /// Conflicts without solutions that need to be solved
+  final List<SyncConflict<T1, T2>> conflicts;
+}
+
+/// Executes the actions required to solve the difference
+Future executeSyncDiff<T1, T2>(
+  final SyncSources<T1, T2> sources,
+  final SyncJournal journal,
+  final SyncDiff<T1, T2> diff,
+) async {
+  for (final action in diff.actions) {
+    switch (action) {
+      case SyncActionDeleteFromA():
+        await sources.sourceA.deleteObject(action.object as SyncObject<T1>);
+        journal.entries.removeWhere((final entry) => entry.id == action.object.id);
```

Comment:
In theory we don't need to add the comparison as the `SyncJournalEntry` already overrides the euqality check by checking the id

```suggestion
        journal.entries.remove(action.object);
```

Reply:
But `action.object.data` is not a `SyncJournalEntry`, that's why the explicit comparison is needed.